python/foglamp/services/core/api/filters.py

```diff
@@ -0,0 +1,270 @@
+__copyright__ = "Copyright (c) 2018 OSIsoft, LLC"
+__license__ = "Apache 2.0"
+__version__ = "${VERSION}"
+
+_help = """
+    ---------------------------------------------------------------------------
+    | POST            | /foglamp/filter                                       |
+    | PUT             | /foglamp/filter/{service_name}/pipeline               |
+    ---------------------------------------------------------------------------
+"""
+
+_LOGGER = logger.setup("filter")
+
+
+async def create_filter(request):
+    """
+    Create a new filter with a specific plugin
+    
+    :Example:
+     curl -X POST http://localhost:8081/foglamp/filter -d 
+     '{
+        "name": "North_Readings_to_PI_scale_stage_1Filter",
+        "plugin": "scale"
+     }'
+
+    'name' is the filter name
+    'plugin' is the filter plugin name
+
+    The plugin is loaded and default config from 'plugin_info'
+    is fetched.
+
+    A new config category 'name' is created:
+    items are:
+       - 'plugin'
+       - all items from default plugin config
+
+    NOTE: The 'create_category' call is made with keep_original_items = True
+
+    """
+    try:
+        # Get inpout data
+        data = await request.json()
+        # Get filter name
+        filter_name = data.get('name', None)
+        # Get plugin name
+        plugin_name = data.get('plugin', None)
+
+        # Check we have needed input data
+        if not filter_name or not plugin_name:
+            raise web.HTTPBadRequest(reason='Filter name or plugin name are required.')
+
+        # Set filter description
+        filter_desc = 'Configuration of \'' + filter_name + '\' filter for plugin \'' + plugin_name + '\''
+        # Get configuration manager instance
+        cf_mgr = ConfigurationManager(connect.get_storage_async())
+        # Load the specified plugin and get plugin data
+        loaded_plugin_info = apiutils.get_plugin_info(plugin_name)
+        # Get plugin default configuration (dict)
+        plugin_config = loaded_plugin_info['config']
+        # Get plugin type (string)
+        loaded_plugin_type = loaded_plugin_info['type']
+        # Get plugin name (string)
+        loaded_plugin_name = plugin_config['plugin']['default']
+
+        # Check first whether filter name already exists
+        category_info = await cf_mgr.get_category_all_items(category_name=filter_name)
+        if category_info is not None:
+            # Filter name already exists: return error
+            message = "Filter '%s' already exists." % filter_name
+            return web.HTTPBadRequest(reason=message)
+
+        # Sanity checks
+        if plugin_name != loaded_plugin_name or loaded_plugin_type != 'filter':
+            error_message = "Loaded plugin '" + loaded_plugin_name + \
+                            "', type '" + loaded_plugin_type + \
+                            "', doesn't match the specified one '" + \
+                            plugin_name + "', type 'filter'"
```

Comment:
Use format string for error_message

Reply:
Is it really needed?

Reply:
@ashish-jabble it looks like this never ending review is done